server/app/auth/oidc/applicant/OidcApplicantProfileAdapter.java

```diff
@@ -0,0 +1,129 @@
+import repository.UserRepository;
+
+/**
+ * This class ensures that the OidcProfileCreator that both the AD and IDCS clients use will
+ * generate a CiviFormProfile object. This is necessary for merging those accounts with existing
+ * accounts - that's not usually needed in web applications which is why we have to write this class
+ * - pac4j doesn't come with it. It's abstract because AD and IDCS need slightly different
+ * implementations of the two abstract methods.
+ */
+public abstract class OidcApplicantProfileAdapter extends OidcProfileAdapter {
+  private final String emailAttributeName;
+  private final Optional<String> localeAttributeName;
+  private final ImmutableList<String> nameAttributeNames;
+
+  public OidcApplicantProfileAdapter(
+      OidcConfiguration configuration,
+      OidcClient client,
+      ProfileFactory profileFactory,
+      Provider<UserRepository> applicantRepositoryProvider,
+      String emailAttributeName,
+      @Nullable String localeAttributeName,
+      ImmutableList<String> nameAttributeNames) {
+    super(configuration, client, profileFactory, applicantRepositoryProvider);
+    this.emailAttributeName = Preconditions.checkNotNull(emailAttributeName);
+    this.localeAttributeName = Optional.ofNullable(localeAttributeName);
+    this.nameAttributeNames = Preconditions.checkNotNull(nameAttributeNames);
+  }
+
+  private Optional<String> getName(OidcProfile oidcProfile) {
+    String name =
+        nameAttributeNames.stream()
+            .filter(s -> !s.isBlank())
+            .map((String attrName) -> oidcProfile.getAttribute(attrName, String.class))
+            .filter(s -> !Strings.isNullOrEmpty(s))
+            .collect(Collectors.joining(" "));
+
+    return Optional.ofNullable(Strings.emptyToNull(name));
+  }
+
+  private Optional<String> getLocale(OidcProfile oidcProfile) {
+    return localeAttributeName
+        .filter(s -> !s.isBlank())
+        .map(name -> oidcProfile.getAttribute(name, String.class))
+        .filter(s -> !Strings.isNullOrEmpty(s));
+  }
+
+  @Override
+  protected String emailAttributeName() {
+    return emailAttributeName;
+  }
+
+  /** Create a totally new Applicant CiviForm profile informed by the provided OidcProfile. */
+  @Override
+  public CiviFormProfile createEmptyCiviFormProfile(OidcProfile profile) {
+    return profileFactory.wrapProfileData(profileFactory.createNewApplicant());
+  }
+
+  protected boolean isTrustedIntermediary(CiviFormProfile profile) {
+    return profile.getAccount().join().getMemberOfGroup().isPresent();
+  }
+
+  @Override
+  protected ImmutableSet<Roles> roles(CiviFormProfile profile, OidcProfile oidcProfile) {
+    if (isTrustedIntermediary(profile)) {
+      return ImmutableSet.of(Roles.ROLE_APPLICANT, Roles.ROLE_TI);
+    }
+    return ImmutableSet.of(Roles.ROLE_APPLICANT);
+  }
+
+  @Override
+  protected void adaptForRole(CiviFormProfile profile, ImmutableSet<Roles> roles) {
+    // not needed
```

Comment:
Not blocking: Consider either removing the comment entirely or expanding it to indicate why a no-op implementation is sufficient.